ESQL: drop `SortAgnostic` interface

bpintea opened the pull request:
This attempts a simplification, dropping an interface: `SortAgnostic` 
* really targets just one rule, `PruneRedundantOrderBy` (needed due to the current incapacity to execute unbounded sorts; the rule doesn't actually solve the issue, but it does alleviate it in some cases).
* has a contract which is in part inconsistent (`STATS` and `SORT` don't need to be in a pattern to have preceding `SORT`s dropped -- https://github.com/elastic/elasticsearch/pull/132417/files#r2285043073);
* it requires cognitive load when adding new commands: this interface is implemented by all commands, except one (`LIMIT`).

bpintea committed:
drop SortAgnostic interface

fang-xing-esql reviewed (approved):
LGTM!

luigidellaquila reviewed (commented):
Thanks @bpintea, I understand how `SortAgnostic` can add some cognitive load, but that's someway on purpose, to save us from functional bugs (wrong results), that are much more complex to spot than errors/exceptions, and IMHO much more serious and problematic.

~~I think the immediate proof of the above is that this PR misses FORK command, that _is not SortAgnostic_ on purpose. With this change, a query like~~

```
from employees 
| SORT emp_no ASC 
| FORK (limit 1) (limit 1)
| SORT emp_no DESC
```

~~will return wrong results after the optimization.~~

~~We can fix it and just add an IF for FORK to the rule, but as soon as we add a new command with similar behavior we'll start silently returning wrong results.~~

[edit] sorry for the noise, limits internal to FORK are enough to let the optimization behave correctly.

Up to you, but IMHO it's risky

luigidellaquila commented:
I had a look at the history of the original PR, and I think this comment, now removed, explains a bit the situation

https://github.com/elastic/elasticsearch/pull/121156/commits/319cc69412df2d41588c1ec9af41d3f3793905a8#diff-7b7a3195c0cfd98ca1ceebc818b4d8a3468b97aa652378b67f3bf406c25c7f5dL95-L98

```
                // IMPORTANT
                // If we introduce window functions or order-sensitive aggs (eg. STREAMSTATS),
                // the previous sort could actually become relevant
                // so we have to be careful with plans that could use them, ie. the following
```

I think we still don't have any of these, the thing that is closer to this is https://github.com/elastic/elasticsearch/pull/137408, but I think it always includes an explicit sort column in the function signature (ie. it never relies on previous order), so we should be safe.

I checked your code more in depth and tested FORK, I can confirm that it woks fine.
ChangePoint was another candidate, but it introduces a SORT+LIMIT, so we are safe also with that.

Again, long term this is risky, but for now I think it works.

alex-spies reviewed (changes requested):
Happy to discuss simplifications, but `SortAgnostic` is actually important because it summarizes, succinctly, what a command needs to do to be compatible with `PruneRedundantOrderBy`. If we remove this, new commands will run into bugs, and we'll have to figure this out again, from scratch.

We're past the time where we can just say "this rule works for all existing commands except this one" because new commands are being added by non-planner-experts. The cognitive load to check if a new command is fine with having sorts in front of it removed is work that needs to be put in to ensure that the optimization doesn't break. Or worse: doesn't _silently_ break, producing query plans that are incorrect without notifying the user (or us, by lack of tests of new commands).

bpintea committed:
add test on all LogicalPlans

bpintea committed:
Merge remote-tracking branch 'upstream/main' into enh/drop_sort_agnostic

bpintea commented:
> I think this comment, now removed, explains a bit the situation

But as the code is pre-this PR, the concerns expressed there have been waved already: `Aggregate` and `Filter` are currently `SortAgnostic`. (Not sure about the inclusion of `Eval` there though.) So...

> the thing that is closer to this is https://github.com/elastic/elasticsearch/pull/137408, but I think it always includes an explicit sort column in the function signature

... even if the functions didn't require a sort criterion, i.e., the functions relied on the existing data ordering, `SortAgnostic` would still not help.

> long term this is risky

Basically, `SortAgnostic` works now as a "simple" warning: "be mindful that there is this rule, PruneRedundantOrderBy, that drops SORTs in certain patterns". Hopefully, the author of a new command (only) would check its significance and possibly add it to their new command. This "watch-out!" could be a test too, which would check all existing LogicalPlans against a list of "this class is safe w.r.t. PruneRedundantOrderBy" and might be safer against overseeing the interface.
Because one could not implement it _and_ not write the right tests. Hopefully a reviewer would catch it.

I've implemented a possible test in [testAllLogicalPlanTypesAreConsidered()](https://github.com/elastic/elasticsearch/pull/137543/files#diff-92b2b661404e647b0d14884daff5d4b666e17c994a02f3cae08f44303ed0dd7fR81).

wchaparro commented:
<!-- analytics-pr-review-management:stall-warn -->

Hi @bpintea — this PR has been identified as **stalled** (open and without meaningful updates for 30+ days).

Please address the requested changes when you can.

It will be **closed in two weeks** if there is no further activity. Please reply, push an update, or close it yourself if it is no longer needed.

— Analytics PR review management

wchaparro commented:
<!-- analytics-pr-review-management:stall-close -->

Hi @bpintea — this PR was marked **stalled** and there has been no activity during the two-week grace period, so it is being **closed**.

Please reopen and update if you still intend to land this change.

— Analytics PR review management
